xdsl/dialects/hw.py

```diff
@@ -409,12 +426,437 @@ def print_parameters(self, printer: Printer):
+            return Direction.parse(parser)
+
+    def print_parameter(self, printer: Printer) -> None:
+        with printer.in_angle_brackets():
+            self.data.print(printer)
+
+
+@irdl_attr_definition
+class ModulePort(ParametrizedAttribute):
+    """
+    Represents a ModulePort. This attribute does not exist in CIRCT
+    but is useful in xDSL to give structure to ModuleType.
+    """
+
+    name = "hw.modport"
+
+    port_name: ParameterDef[StringAttr]
+    type: ParameterDef[TypeAttribute]
+    dir: ParameterDef[DirectionAttr]
+
+
+@irdl_attr_definition
+class ModuleType(ParametrizedAttribute, TypeAttribute):
+    name = "hw.modty"
+
+    ports: ParameterDef[ArrayAttr[ModulePort]]
+
+    @classmethod
+    def parse_parameters(cls, parser: AttrParser) -> Sequence[Attribute]:
+        def parse_port() -> ModulePort:
+            direction = Direction.parse(parser)
+            if (
+                name := parser.parse_optional_identifier()
+                or parser.parse_optional_str_literal()
+            ) is None:
+                parser.raise_error(
+                    "expected port name as identifier or string litteral"
+                )
+
+            parser.parse_punctuation(":")
+            typ = parser.parse_type()
+
+            return ModulePort([StringAttr(name), typ, DirectionAttr(direction)])
+
+        return [
+            ArrayAttr(
+                parser.parse_comma_separated_list(parser.Delimiter.ANGLE, parse_port)
+            )
+        ]
+
+    def print_parameters(self, printer: Printer):
+        def print_port(port: ModulePort):
+            port.dir.data.print(printer)
+            printer.print(" ")
+            printer.print_identifier_or_string_literal(port.port_name.data)
+            printer.print(" : ")
+            printer.print_attribute(port.type)
+
+        with printer.in_angle_brackets():
+            printer.print_list(self.ports.data, print_port)
+
+
+@irdl_attr_definition
+class ParamDeclAttr(ParametrizedAttribute):
+    name = "hw.param.decl"
+
+    port_name: ParameterDef[StringAttr]
+    type: ParameterDef[TypeAttribute]
+
+    @classmethod
+    def parse_free_standing_parameters(
+        cls, parser: AttrParser, always_parse_name_in_quotes: bool = False
+    ) -> Sequence[Attribute]:
+        """
+        Parses the parameter declaration without the encompassing angle brackets.
+        """
+
+        name = parser.parse_optional_str_literal()
+        if name is None and not always_parse_name_in_quotes:
+            name = parser.parse_optional_identifier()
+        if name is None:
+            parser.raise_error("expected parameter name")
+        parser.parse_punctuation(":")
+        typ = parser.parse_attribute()
+        if not isinstance(typ, TypeAttribute):
+            parser.raise_error("expected type attribute for parameter")
+
+        if parser.parse_optional_punctuation("=") is not None:
+            # TODO: support default values for parameters
```

Comment:
Should this be an issue too? CC: @superlopuh

Reply:
#2367